### PR TITLE
core(network-request): add new timing properties

### PR DESCRIPTION
**Pull request by connorjclark**
split off from #14311

new time properties / renames
```
Renamed:

startTime            -> networkRequestTime
endTime              -> networkEndTime
responseReceivedTime -> responseHeadersEndTime

Added:

rendererStartTime
rendererEndTime 
rendererEndTime is not actually something we have yet (only exists in trace), so at the moment it's always equal to networkEndTime. It is included to fill out our model of the network, and in some places the former is used instead of the latter even though it doesn't make a difference yet. At the very least, it better documents intent.
```

startTime/endTime/responseReceivedTime have stuck around as aliases, as to keep this PR small (it'd be dozens of files with just renames) and focused.

excerpt from network request constructor:

```
/** When the renderer process initially discovers a network request, in milliseconds. */
this.rendererStartTime = -1;
/**
 * When the network service is about to handle a request, ie. just before going to the
 * HTTP cache or going to the network for DNS/connection setup, in milliseconds.
 */
this.networkRequestTime = -1;
/** When the last byte of the response headers is received, in milliseconds. */
this.responseHeadersEndTime = -1;
/** When the last byte of the response body is received, in milliseconds. */
this.networkEndTime = -1;
/**
 * Network response is done, this marks the time when the main thread is first available
 * to actually utilize the network resource, in milliseconds.
 *
 * TODO: we don't have this on the CDP, but it exists in trace as ResourceFinish.ts
 * So currently this is equivalent to networkEndTime.
 */
this.rendererEndTime = -1;
```

### CODE REVIEW

File: core/test/network-records-to-devtools-log.js [hunk 1]
```diff
@@ -91,19 +91,36 @@ function getRequestServedFromCacheEvent(networkRecord, index) {
...
   const headers = headersArrayToHeadersDict(networkRecord.responseHeaders);
-  let timing;
-  if (networkRecord.timing) {
-    timing = {...networkRecord.timing};
-    if (timing.requestTime === undefined) {
-      timing.requestTime = networkRecord.startTime / 1000 || 0;
-    }
-  }
+  // let timing;
+  // if (networkRecord.timing) {
```
Comment (connorjclark) [inline core/test/network-records-to-devtools-log.js:95]: oops, forgot to remove.

Note that this code has moved. I tried to explain in the comment elsewhere.

File: core/audits/network-requests.js [hunk 2]
```diff
@@ -64,8 +64,11 @@ class NetworkRequests extends Audit {
       return {
         url: UrlUtils.elideDataURI(record.url),
         protocol: record.protocol,
-        startTime: normalizeTime(record.startTime),
-        endTime: normalizeTime(record.endTime),
+        rendererStartTime: normalizeTime(record.rendererStartTime),
+        networkRequestTime: normalizeTime(record.networkRequestTime),
+        responseHeadersEndTime: normalizeTime(record.responseHeadersEndTime),
+        networkEndTime: normalizeTime(record.networkEndTime),
+        rendererEndTime: normalizeTime(record.rendererEndTime),
```
Comment (brendankenny) [inline core/audits/network-requests.js:71]: adding the extra new properties could possibly wait until all the other audits are switched over to using them (where appropriate). That would keep those changes separate from the renames.

File: core/lib/network-request.js [hunk 3]
```diff
@@ -184,6 +195,50 @@ class NetworkRequest {
     this.isLinkPreload = false;
   }
 
+  // TODO: remove after timing refactor is done
+  // See https://github.com/GoogleChrome/lighthouse/pull/14311
+  /**
```
Comment (brendankenny) [inline core/lib/network-request.js:200]: Easier just to `s/startTime/networkRequestTime/g`, `s/endTime/networkEndTime/g` and `s/responseReceivedTime/responseHeadersEndTime/g`? Seems unnecessary to have the codebase use both (and necessitates the `network-records-to-devtools-log.js` workaround) when there's very little review work for just a property rename.

Comment (connorjclark) [reply core/lib/network-request.js:200]: The big reason for this approach is that this PR is just 7 files (less, if I take your suggestion to defer network requests audit changes)- but it would be dozens of files mixed in with the important changes here if done all at once.

The few lines in records to DevTools log is just encoding the framework we'd mostly use during the conversion process (next PR).

Comment (brendankenny) [reply core/lib/network-request.js:200]: I was probably too focused on an exact approach in my suggestion :) I think we're all broadly going for the same thing, my main issue was just anticipating a PR where some `startTime`s get changed to `networkRequestTime` and some get changed to `rendererStartTime` (and similar for `endTime`) and a reviewer will have to read each line carefully to double check which one was used.

I was attempting to suggest
- a PR that only updates `startTime` to `networkRequestTime` (and `endTime` to  `networkEndTime`) so no actual timing/test values changed, and then
- a PR that changes some of those `networkRequestTime`s to `rendererStartTime` (and `networkEndTime` to `rendererEndTime`), which would involve some timing/test value changes. That way the changes we want to be most deliberate about are highlighted (any property being changed is switched to the new timing value).

If intermediate PRs before that point help break it up nicely, SGTM, I won't object to that :)

File: core/audits/network-requests.js [hunk 4]
```diff
@@ -111,8 +118,8 @@ class NetworkRequests extends Audit {
     const tableDetails = Audit.makeTableDetails(headings, results);
 
     // Include starting timestamp to allow syncing requests with navStart/metric timestamps.
-    const networkStartTimeTs = Number.isFinite(earliestStartTime) ?
-        earliestStartTime * 1000 : undefined;
+    const networkStartTimeTs = Number.isFinite(earliestMainThreadStartTime) ?
+        earliestMainThreadStartTime * 1000 : undefined;
```
[the remark this answers is not in the document]
Comment (brendankenny) [reply core/audits/network-requests.js:122]: > i'm not sure about this `* 1000`, given that rendererstartime should be in ms.

Yeah, I guess it's debatable what it should use. My original thinking was it should use microseconds because you compare it against `observedNavigationStartTs` to get the offset between CPU task timestamps and network times:

```sql
(CAST(JSON_VALUE(lighthouse, '$.audits."network-requests".details.debugData.networkStartTimeTs') AS FLOAT64) - CAST(JSON_VALUE(lighthouse, '$.audits.metrics.details.items[0].observedNavigationStartTs') AS FLOAT64)) / 1000 AS network_ts_offset
```

I put the `Ts` at the end of `networkStartTimeTs` to try to communicate that, but it's also a good argument that it's simpler if this audit only had values in ms 🤷

File: core/test/network-records-to-devtools-log.js [hunk 5]
```diff
@@ -227,6 +244,48 @@ function addRedirectResponseIfNeeded(networkRecords, record) {
...
+      // If networkRecord.timing explicitly does not having a timing object, only create one
+      // if we absolutely need it. See _recomputeTimesWithResourceTiming.
+      const netReqTime = networkRecord.networkRequestTime;
+      const willNeedTimingObject =
+        (netReqTime !== undefined && netReqTime !== networkRecord.rendererStartTime) ||
+        (networkRecord.responseHeadersEndTime !== undefined);
+      if (willNeedTimingObject) networkRecord.timing = networkRecord.timing || {};
+
+      if (networkRecord.timing) {
+        if (networkRecord.timing.requestTime === undefined) {
```
Comment (brendankenny) [inline core/test/network-records-to-devtools-log.js:280]: don't these both need to be set if one of them is going to be?

File: core/lib/network-request.js (hunk 3 above)
Comment (connorjclark) [reply core/lib/network-request.js:200]: > a PR that only updates startTime to networkRequestTime (and endTime to networkEndTime) so no actual timing/test values changed

Doing this next makes sense for the reasons you highlighted. Will do that. Hopefully it'll be the fasted on these PRs!

File: core/lib/network-request.js [hunk 6]
```diff
@@ -366,28 +425,32 @@ class NetworkRequest {
...
-    this.responseReceivedTime = Math.max(this.responseReceivedTime, this.startTime);
+    this.responseHeadersEndTime =
+      Math.min(this.responseHeadersEndTime, headersReceivedTime);
+    this.responseHeadersEndTime =
+      Math.max(this.responseHeadersEndTime, this.networkRequestTime);
     // We're only at responseReceived (_onResponse) at this point.
     // This endTime may be redefined again after onLoading is done.
-    this.endTime = Math.max(this.endTime, this.responseReceivedTime);
+    this.networkEndTime = Math.max(this.networkEndTime, this.responseHeadersEndTime);
+    this.rendererEndTime = this.networkEndTime;
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/lib/network-request.js:444]: Good point. I am simply matching previous behavior here, so I think we should change the endTime/networkEndTime in isolation. But I could delete `rendererEndTime` here for now.

File: core/test/network-records-to-devtools-log.js (hunk 5 above)
Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js:280]: Yeah, but the idea here was to never override a provided timing object, in case a test defined one (even partially).

Maybe instead this should check if both these timing fields are undefined before assigning either?

File: core/test/network-records-to-devtools-log.js [hunk 7]
```diff
@@ -227,6 +244,48 @@ function addRedirectResponseIfNeeded(networkRecords, record) {
...
+        networkRecord.rendererStartTime = networkRecord.startTime;
+      }
+      if (networkRecord.endTime !== undefined) {
+        networkRecord.networkEndTime = networkRecord.endTime;
+      }
+      if (networkRecord.responseReceivedTime !== undefined) {
+        networkRecord.responseHeadersEndTime = networkRecord.responseReceivedTime;
+      }
+
+      // Set timing.requestTime and timing.receiveHeadersEnd to be values that
```
Comment (connorjclark) [inline core/test/network-records-to-devtools-log.js:266]: I thought I could add a precondition here:

```js
if (record.rendererStartTime === undefined ^ record.networkEndTime === undefined) {
        // eslint-disable-next-line max-len
        throw new Error('expected test record to define both rendererStartTime and networkEndTime, but never just one');
      }
```

But 10 tests threw errors because they mocked network records and only defined startTime (or endTime). Should I update those tests so we can ensure to more realistic values here?

Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js:266]: actually all are from uses-rel-preload, and all just set startTime leaving endTime undefined.

it'd be easy to set endTime to something in those tests (-1, I guess that is the implicit default getting used?)

or maybe assign endTime -1 if startTime is defined and endTime is missing on a test record object?

File: core/test/network-records-to-devtools-log.js [hunk 8]
```diff
@@ -227,6 +244,48 @@ function addRedirectResponseIfNeeded(networkRecords, record) {
...
+      // so this roundtrips correctly. Unless, of course, timing values are explicitly set
+      // already.
+
+      // If networkRecord.timing explicitly does not having a timing object, only create one
+      // if we absolutely need it. See _recomputeTimesWithResourceTiming.
+      const netReqTime = networkRecord.networkRequestTime;
+      const willNeedTimingObject =
+        (netReqTime !== undefined && netReqTime !== networkRecord.rendererStartTime) ||
+        (networkRecord.responseHeadersEndTime !== undefined);
+      if (willNeedTimingObject) networkRecord.timing = networkRecord.timing || {};
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js:277]: ugh, the problem here is that `networkRequestTime` isn't properly set when `timing` is missing, failing 8 tests in `network-request-test.js`.

For example, take `update fetch stats for Lightrider` test. the `startTime`/`networkRequestTime` value is now 50 instead of 0, because I wanted some separation between the first two timing properties. Without a timing object, such a distinction can never be made.

Should I 1) fix the tests to provide the relevant timing object to make the values come out right or 2) do a check in this records->log conversion if these properties differ, and only create a timing object magically...to reduce some test boilerplate. Leaning towards 1.

File: core/lib/network-request.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply core/lib/network-request.js:444]: yeah, just a comment on the existing code. Fix here or later, up to you. It's also not harming anything, just weird.

File: core/test/network-records-to-devtools-log.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply core/test/network-records-to-devtools-log.js]: maybe it would be easier to always create a timing object now?

[the remark this answers is not in the document]
Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js]: fwiw in the correct tests `willNeedTimingObject` is always false for uses-rel-preconnect ... and there are zero tests in that audit that exercise a missing timing object. Is this a valid scenario?

File: core/test/network-records-to-devtools-log.js [hunk 9]
```diff
@@ -225,28 +242,87 @@ function addRedirectResponseIfNeeded(networkRecords, record) {
  * @return {LH.DevtoolsLog}
  */
 function networkRecordsToDevtoolsLog(networkRecords, options = {}) {
+  // Clone test network records objects before potential modifications.
+  networkRecords = networkRecords.map(record => {
+    if (record.constructor === Object) {
```
Comment (connorjclark) [inline core/test/network-records-to-devtools-log.js:247]: I can remove this object literal check and the one below and tests still all pass. I believe only the `'should roundtrip a real devtools log properly'` test actual passes in real NetworkRequest instances. It seems wrong to modify those though, hence the checks here.

File: core/test/audits/network-requests-test.js [hunk 10]
```diff
@@ -62,14 +62,17 @@ describe('Network requests audit', () => {
...
 
   it('should handle times correctly', async () => {
     const records = [
-      {url: 'https://example.com/0', startTime: 15.0, endTime: 15.5},
-      {url: 'https://example.com/1', startTime: 15.5, endTime: -1},
+      // Note: should use rendererEndTime but that is currently a "dummy" value equivalent
+      // to networkEndTime, and our networkRecordsToDevtoolsLog has no ability to roundtrip
+      // that.
+      {url: 'https://example.com/0', rendererStartTime: 15.0, networkEndTime: 15.5},
+      {url: 'https://example.com/1', rendererStartTime: 15.5, networkEndTime: -1},
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/test/audits/network-requests-test.js:84]: done.

but why is the second item here considered `finished`?

File: core/test/network-records-to-devtools-log.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js]: bump for the above question

File: core/test/network-records-to-devtools-log.js (hunk 7 above)
Comment (brendankenny) [reply core/test/network-records-to-devtools-log.js:282]: it's fine to update the tests, but we do want some solution in here for future test writers, either picking a reasonable value or cause some failure (e.g. in the roundtripping) so the test author knows something is wrong

File: core/test/network-records-to-devtools-log.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply core/test/network-records-to-devtools-log.js]: > fwiw for current PR code, willNeedTimingObject is always false for uses-rel-preconnect ... and there are zero tests in that audit that exercise a missing timing object. Is this a valid scenario?

Maybe just for failed requests? I'm not sure if it's important, it's possible it's just inherited from super old SDK code, or it could handle some important case we're not thinking of :)

**Comment by brendankenny** [comment]
also, FWIW, there's a `network-records-to-devtools-log-test.js` if that would be helpful to cement some of these corner cases

File: core/test/network-records-to-devtools-log.js (hunk 7 above)
Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js:258]: I thought this would work:

```
if (record.rendererStartTime === undefined) {
        record.rendererStartTime = 0;
      }
      if (record.networkEndTime === undefined) {
        record.networkEndTime = -1;
      }
```

But had complications getting it to work w/ the `redirectResponseTimestamp` kludge used elsewhere. I'm getting a bit off track here, so gonna forget about normalizing the test data for now.

File: core/test/network-records-to-devtools-log.js [hunk 11]
```diff
@@ -225,28 +242,87 @@ function addRedirectResponseIfNeeded(networkRecords, record) {
...
+      // so this roundtrips correctly. Unless, of course, timing values are explicitly set
+      // already.
+
+      // If record.timing explicitly does not having a timing object, only create one
+      // if we absolutely need it. See _recomputeTimesWithResourceTiming.
+      const netReqTime = record.networkRequestTime;
+      const willNeedTimingObject =
+        (netReqTime !== undefined && netReqTime !== record.rendererStartTime) ||
+        (record.responseHeadersEndTime !== undefined);
+      if (willNeedTimingObject) record.timing = record.timing || {};
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js]: Ah, I recall why I did it this way. See the test failure for `yarn mocha gatherers/inspector-issues`:

```
      "sessionId": undefined,
      "statusCode": -1,
-     "timing": undefined,
+     "timing": Object {
+       "receiveHeadersEnd": 0,
+       "requestTime": -0.001,
+     },
      "transferSize": 0,
      "url": "https://example.com",
    },
  ]
      at networkRecordsToDevtoolsLog (file:///Users/cjamcl/src/lighthouse/core/test/network-records-to-devtools-log.js:295:40)
      at Context.<anonymous> (file:///Users/cjamcl/src/lighthouse/core/test/gather/gatherers/inspector-issues-test.js:329:19)
      at processImmediate (node:internal/timers:466:21)
```

Since timing object presence is now being automatically forced, the roundtrip assertion trips up unless we do the modification on the (cloned) test record.

So uh.... should I put it back?

[the remark this answers is not in the document]
Comment (brendankenny) [reply core/test/network-records-to-devtools-log.js]: Part of it is `inspector-issues-test` is weird: it's `Object.assign`ing to ` real `NetworkRequest`, then passing it through `networkRecordsToDevtoolsLog` (it could have just used the partial passed through `networkRecordsToDevtoolsLog`, but it looks like those lines came a long time after the initial `mockRequest` method was written).

But even fixing that, there's a deeper problem and I don't think we can get away with optionally adding `timing` anymore. If you add

```diff
diff --git a/core/test/network-records-to-devtools-log-test.js b/core/test/network-records-to-devtools-log-test.js
index d114d483b..dd176e057 100644
--- a/core/test/network-records-to-devtools-log-test.js
+++ b/core/test/network-records-to-devtools-log-test.js
@@ -20,6 +20,11 @@ describe('networkRecordsToDevtoolsLog', () => {
       {method: 'Network.dataReceived'},
       {method: 'Network.loadingFinished'},
     ]);
+
+    const roundTripRecords = NetworkRecorder.recordsFromLogs(log);
+    const logNo2 = networkRecordsToDevtoolsLog(roundTripRecords);
+    const roundTripRecordsNo2 = NetworkRecorder.recordsFromLogs(logNo2);
+    expect(roundTripRecordsNo2).toMatchObject(roundTripRecords);
   });
 
   it('should set resource and transfer sizes', () => {
```
and run `yarn mocha network-records-to-devtools-log` this fails on this branch but passes on `main`. The first round-trip check (inside `networkRecordsToDevtoolsLog`) passes, but running the output through it again it fails on the same `timing` `undefined` vs `Object` part (the first time through it doesn't add a `timing` object, but the second time it does)

File: core/test/network-records-to-devtools-log.js (hunk 1 above)
Comment (Waazer) [reply core/test/network-records-to-devtools-log.js]: Kompitabel korleis type i dag

File: core/audits/network-requests.js (hunk 2 above)
Comment (Waazer) [reply core/audits/network-requests.js:71]: Kom bedre å bare ta det litt av Kombef er litt for tynn Henrik
